internal/extender/resourcereservations.go

```diff
@@ -15,26 +15,281 @@
+		rrm.softReservationStore.CreateSoftReservationIfNotExists(driver.Labels[common.SparkAppIDLabel])
+	}
+
+	return rr, nil
+}
+
+// FindAlreadyBoundReservationNode returns a node name that was previously allocated to this executor, if any.
+// Binding reservations have to be idempotent. Binding the pod to the node on kube-scheduler might fail, so we want to get the same executor pod on a retry.
+func (rrm *ResourceReservationManager) FindAlreadyBoundReservationNode(ctx context.Context, executor *v1.Pod) (string, error) {
+	resourceReservation, ok := rrm.GetResourceReservation(executor)
+	if !ok {
+		return "", werror.ErrorWithContextParams(ctx, "failed to get resource reservations")
+	}
+	for name := range resourceReservation.Spec.Reservations {
+		if resourceReservation.Status.Pods[name] == executor.Name {
+			return resourceReservation.Spec.Reservations[name].Node, nil
+		}
+	}
+
+	// Check any soft reservations as well
+	if sr, ok := rrm.softReservationStore.GetExecutorSoftReservation(ctx, executor); ok {
+		return sr.Node, nil
+	}
+	return "", nil
+}
+
+// FindUnboundReservationNodes returns a slice of node names that have unbound reservations for this Spark application.
+// This includes both reservations we have not yet scheduled any executors on as well as reservations that have executors that are now dead.
+// Spark will recreate lost executors, so the replacement executors should be placed on the reserved spaces of dead executors.
+func (rrm *ResourceReservationManager) FindUnboundReservationNodes(ctx context.Context, executor *v1.Pod) ([]string, error) {
+	unboundReservationsToNodes, err := rrm.getUnboundReservations(ctx, executor)
+	if err != nil {
+		return []string{}, err
+	}
+	unboundReservationNodes := utils.NewStringSet(len(unboundReservationsToNodes))
+	for _, node := range unboundReservationsToNodes {
+		unboundReservationNodes.Add(node)
+	}
+	return unboundReservationNodes.ToSlice(), nil
+}
+
+// GetFreeExecutorSpots returns the number of executors the application can still schedule.
+func (rrm *ResourceReservationManager) GetFreeExecutorSpots(ctx context.Context, executor *v1.Pod) (int, error) {
+	unboundReservations, err := rrm.getUnboundReservations(ctx, executor)
+	if err != nil {
+		return 0, err
+	}
+	extraExecutorFreeSpots, err := rrm.getFreeExtraExecutorSpots(ctx, executor)
+	if err != nil {
+		return 0, err
+	}
+	return len(unboundReservations) + extraExecutorFreeSpots, nil
+}
+
+// ReserveForExecutor creates a reservation for the passed executor on the passed node. This reservation could either be a resource reservation,
+// or a soft reservation if dynamic allocation is enabled.
+func (rrm *ResourceReservationManager) ReserveForExecutor(ctx context.Context, executor *v1.Pod, node string) error {
+	rrm.mutex.Lock()
+	defer rrm.mutex.Unlock()
+
+	// TODO: make sure executor doesn't already have a reservation, if it does, it must be freed (we don't seem to handle this in the current logic, but we should going forward)
```

Comment:
don't we handle this at `FindAlreadyBoundReservationNode`?

Reply:
We could find an already bound reservation, but for a node that is no longer available. In this case, we will reschedule the executor onto a new node and call `ReserveForExecutor` for this one.
This means that we will double reserve nodes for a single executor.

Reply:
Fixed by counting reservations bound to executors on a node different from the reservation node, as an unbound reservation.